File: Src/ECScape.Core/Entities/EntityFactory.cs
using ECScape.Core.Components;
using ECScape.Core.Engine;

namespace ECScape.Core.Entities;

public static class EntityFactory
{
    private static int idCounter = 0;

    public static Entity CreatePlayer(World world, double left, double top)
    {
        return Create(world, [
            new Exists(),
            new Health(),
            new AffectedByGravity(),
            new LimitedByBounds(),
            new Position(left, top),
            new Drawable { Symbol = '☻', Color = ConsoleColor.Green },
            new Size(3, 3),
            new Velocity(),
            new PlayerControllable(),
            new Statistics(),
            new Invulnerable
            {
                ExpirationTime = DateTime.Now.AddSeconds(3)
            },
        ]);
    }

    public static Entity CreateEnemy(World world, double left, double top, char symbol, int width, int height)
    {
        return Create(world, [
            new Exists(),
            new AffectedByGravity(),
            new LimitedByBounds(),
            new Position(left, top),
            new Npc(),
            new Drawable { Symbol = symbol, Color = ConsoleColor.Red },
            new Size(width, height),
            new Velocity(),
            new DamagesPlayer { DamageAmount = 1 }
        ]);
    }

    public static Entity CreateMeteoroid(World world, double left, int width, int height, bool spawner)
    {
        var entity = Create(world, [
            new AffectedByGravity(),
            new Position(left, UiInterface.WorldTop),
            new Drawable { Symbol = '★', Color = ConsoleColor.DarkRed },
            new Size(width, height),
            new Velocity(){ X = World.Random.Next(-2,2), Y = 5 },
            new DamagesPlayer { DamageAmount = 1 }
        ]);

        if (spawner)
        {
            entity.AddComponent(new Spawner { LastSpawn = DateTime.Now, RandomSizeHealth = false });
        }
        else
        {
            entity.AddComponent(new Exists());
        }

        return entity;
    }

    public static Entity CreatePowerUp(World world, double left, int width, int height, bool spawner)
    {
        var entity = Create(world, [
            new AffectedByGravity() { Gravity = 3 },
            new PowerUpHealth(),
            new Position(left, UiInterface.WorldTop),
            new Drawable { Symbol = '+', Color = ConsoleColor.Cyan },
            new Size(width, height),
            new Velocity() { X = World.Random.Next(-2,2), Y = 0 }
        ]);

        if (spawner)
        {
            entity.AddComponent(new Spawner { LastSpawn = DateTime.Now, RandomSizeHealth = false });
        }
        else
        {
            entity.AddComponent(new Exists());
        }

        return entity;
    }

    public static Entity CreateCollectable(World world, double left, double top, char symbol, int width, int height, int hearts, bool spawner)
    {
        var entity = Create(world, [
            new Health { Hearts = hearts },
            new AffectedByGravity(),
            new LimitedByBounds(),
            new Position(left, top),
            new Npc(),
            new Drawable { Symbol = symbol, Color = ConsoleColor.Yellow },
            new Size(width, height),
            new Velocity(),
            new Collectable(),
            new Invulnerable
            {
                ExpirationTime = DateTime.Now.AddSeconds(3)
            },
        ]);

        if (spawner)
        {
            entity.AddComponent(new Spawner());
        }
        else
        {
            entity.AddComponent(new Exists());
        }

        return entity;
    }

    public static Entity CreateGroundBlock(World world, int left, int top, int width, int height)
    {
        return Create(world, [
            new Exists(),
            new Position(left, top),
            new Drawable { Symbol = '=', Color = ConsoleColor.Gray },
            new Size(width, height),
            new LimitedByBounds(),
            new Solid()
        ]);
    }

    public static Entity CreateCloudBlock(World world, int left, int top, int width, int height)
    {
        return Create(world, [
            new Exists(),
            new Position(left, top),
            new Drawable { Symbol = '=', Color = ConsoleColor.Magenta },
            new Size(width, height),
            new LimitedByBounds(),
            new Solid()
        ]);
    }

    private static Entity Create(World world, List<IComponent> components)
    {
        var id = idCounter++;
        var entity = new Entity(id, components);
        world.Entities.Add(entity);
        return entity;
    }
}

File: src/entities/EntityFactory.js
import {
    Exists, Health, AffectedByGravity, LimitedByBounds, Position, Drawable, Size, Velocity, PlayerControllable, Statistics, Invulnerable, Npc, DamagesPlayer,
    PowerUpHealth, Collectable, Solid
} from '../components/Components.js';
import Entity from '../entities/Entity.js';
import UiInterface from '../core/uiInterface.js';

class EntityFactory {
    static idCounter = 0;

    static create(world, components) {
        const id = this.idCounter++;
        const entity = new Entity(id);
        components.forEach(c => entity.addComponent(c));
        world.Entities.push(entity);
        return entity;
    }

    static createPlayer(world, left, top) {
        return this.create(world, [
            new Exists(),
            new Health(),
            new AffectedByGravity(),
            new LimitedByBounds(),
            new Position(left, top),
            new Drawable('?', 'green'),
            new Size(3, 3),
            new Velocity(),
            new PlayerControllable(),
            new Statistics(),
            new Invulnerable({
                ExpirationTime: Date.now() + 3000
            }),
        ]);
    }

    static createEnemy(world, left, top, symbol, width, height) {
        return this.create(world, [
            new Exists(),
            new AffectedByGravity(),
            new LimitedByBounds(),
            new Position(left, top),
            new Npc(),
            new Drawable(symbol, 'red'),
            new Size(width, height),
            new Velocity(),
            new DamagesPlayer(1)
        ]);
    }

    static createMeteoroid(world, left, width, height) {
        return this.create(world, [
            new Exists(),
            new AffectedByGravity(),
            new Position(left, UiInterface.WorldTop),
            new Drawable('?', 'darkred'),
            new Size(width, height),
            new Velocity(
                Math.floor(-2 + Math.random() * 4),
                5
            ),
            new DamagesPlayer(1)
        ]);
    }

    static createPowerUp(world, left, width, height) {
        return this.create(world, [
            new Exists(),
            new AffectedByGravity(3),
            new PowerUpHealth(),
            new Position(left, UiInterface.WorldTop),
            new Drawable('+', 'cyan'),
            new Size(width, height),
            new Velocity(
                Math.floor(-2 + Math.random() * 4),
                0
            )
        ]);
    }

    static createCollectable(world, left, top, symbol, width, height, hearts) {
        return this.create(world, [
            new Exists(),
            new Health(hearts),
            new AffectedByGravity(),
            new LimitedByBounds(),
            new Position(left, top),
            new Npc(),
            new Drawable(symbol, 'yellow'),
            new Size(width, height),
            new Velocity(),
            new Collectable(),
            new Invulnerable({
                ExpirationTime: Date.now() + 3000
            }),
        ]);
    }

    static createGroundBlock(world, left, top) {
        return this.create(world, [
            new Exists(),
            new Position(left, top),
            new Drawable('=', 'gray'),
            new Size(1, 1),
            new LimitedByBounds(),
            new Solid()
        ]);
    }

    static createCloudBlock(world, left, top) {
        return this.create(world, [
            new Exists(),
            new Position(left, top),
            new Drawable('=', 'magenta'),
            new Size(1, 1),
            new LimitedByBounds(),
            new Solid()
        ]);
    }
}

export default EntityFactory;